### PR TITLE
[INFRA/CORE] Refactor and add unit tests to turn code

### PR DESCRIPTION
**Pull request by lhl2617**
# Summary

As title.

## Additional Information

N/A

## Test Plan

Added unit tests! 👯 

- `go test ./...`
```
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]   
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]   
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]   
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]   
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]   
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]   
ok      github.com/SOMAS2020/SOMAS2020/internal/common  (cached) [no tests to run]
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]   
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    (cached)
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]   
ok      github.com/SOMAS2020/SOMAS2020/internal/server  (cached)
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]
```

- `go run .`
```
2020/12/18 13:03:10 [SERVER]: start runTurn
2020/12/18 13:03:10 [SERVER]: TURN: 1, Season: 1
2020/12/18 13:03:10 [SERVER]: start updateIsland
2020/12/18 13:03:10 [Team3]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [Team4]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [Team5]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [Team6]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [Team1]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [Team2]: Received start of turn game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/18 13:03:10 [SERVER]: finish updateIsland
2020/12/18 13:03:10 [SERVER]: start runOrgs
2020/12/18 13:03:10 [SERVER]: start runIIGO
2020/12/18 13:03:10 [SERVER]: finish runIIGO
2020/12/18 13:03:10 [SERVER]: start runIIFO
2020/12/18 13:03:10 [SERVER]: finish runIIFO
2020/12/18 13:03:10 [SERVER]: start runIITO
2020/12/18 13:03:10 [SERVER]: finish runIITO
2020/12/18 13:03:10 [SERVER]: finish runOrgs
2020/12/18 13:03:10 [SERVER]: start getAndDispatchEndOfTurnActions
2020/12/18 13:03:10 [Team3]: EndOfTurnActions
2020/12/18 13:03:10 [Team4]: EndOfTurnActions
2020/12/18 13:03:10 [Team5]: EndOfTurnActions
2020/12/18 13:03:10 [Team6]: EndOfTurnActions
2020/12/18 13:03:10 [Team1]: EndOfTurnActions
2020/12/18 13:03:10 [Team2]: EndOfTurnActions
2020/12/18 13:03:10 [GAMESTATE]: start DispatchActions
2020/12/18 13:03:10 [GAMESTATE]: finish DispatchActions
2020/12/18 13:03:10 [SERVER]: finish getAndDispatchEndOfTurnActions
2020/12/18 13:03:10 [SERVER]: start endOfTurn
2020/12/18 13:03:10 [SERVER]: start probeDisaster
2020/12/18 13:03:10 [SERVER]: finish probeDisaster
2020/12/18 13:03:10 [SERVER]: start incrementTurnAndSeason
2020/12/18 13:03:10 [SERVER]: finish incrementTurnAndSeason
2020/12/18 13:03:10 [SERVER]: start deductCostOfLiving
2020/12/18 13:03:10 [SERVER]: finish deductCostOfLiving
2020/12/18 13:03:10 [SERVER]: start updateIslandLivingStatus
2020/12/18 13:03:10 [SERVER]: finish updateIslandLivingStatus
2020/12/18 13:03:10 [SERVER]: finish endOfTurn
2020/12/18 13:03:10 [SERVER]: finish runTurn
2020/12/18 13:03:11 [SERVER]: start runTurn
2020/12/18 13:03:11 [SERVER]: TURN: 2, Season: 1
2020/12/18 13:03:11 [SERVER]: start updateIsland
2020/12/18 13:03:11 [Team4]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [Team5]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [Team6]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [Team1]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [Team2]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [Team3]: Received start of turn game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/18 13:03:11 [SERVER]: finish updateIsland
2020/12/18 13:03:11 [SERVER]: start runOrgs
2020/12/18 13:03:11 [SERVER]: start runIIGO
2020/12/18 13:03:11 [SERVER]: finish runIIGO
2020/12/18 13:03:11 [SERVER]: start runIIFO
2020/12/18 13:03:11 [SERVER]: finish runIIFO
2020/12/18 13:03:11 [SERVER]: start runIITO
2020/12/18 13:03:11 [SERVER]: finish runIITO
2020/12/18 13:03:11 [SERVER]: finish runOrgs
2020/12/18 13:03:11 [SERVER]: start getAndDispatchEndOfTurnActions
2020/12/18 13:03:11 [Team4]: EndOfTurnActions
2020/12/18 13:03:11 [Team5]: EndOfTurnActions
2020/12/18 13:03:11 [Team6]: EndOfTurnActions
2020/12/18 13:03:11 [Team1]: EndOfTurnActions
2020/12/18 13:03:11 [Team2]: EndOfTurnActions
2020/12/18 13:03:11 [Team3]: EndOfTurnActions
2020/12/18 13:03:11 [GAMESTATE]: start DispatchActions
2020/12/18 13:03:11 [GAMESTATE]: finish DispatchActions
2020/12/18 13:03:11 [SERVER]: finish getAndDispatchEndOfTurnActions
2020/12/18 13:03:11 [SERVER]: start endOfTurn
2020/12/18 13:03:11 [SERVER]: start probeDisaster
2020/12/18 13:03:11 [SERVER]: finish probeDisaster
2020/12/18 13:03:11 [SERVER]: start incrementTurnAndSeason
2020/12/18 13:03:11 [SERVER]: finish incrementTurnAndSeason
2020/12/18 13:03:11 [SERVER]: start deductCostOfLiving
2020/12/18 13:03:11 [SERVER]: finish deductCostOfLiving
2020/12/18 13:03:11 [SERVER]: start updateIslandLivingStatus
2020/12/18 13:03:11 [SERVER]: finish updateIslandLivingStatus
2020/12/18 13:03:11 [SERVER]: finish endOfTurn
2020/12/18 13:03:11 [SERVER]: finish runTurn
2020/12/18 13:03:11 [SERVER]: Max turns '2' reached or exceeded
===== START OF TURN 1 (END OF TURN 0) =====
common.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
common.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
common.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team5:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
```

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/server/turn.go [hunk 1]
```diff
@@ -101,42 +101,48 @@ func (s *SOMASServer) getAndDispatchEndOfTurnActions() error {
...
 // deductCostOfLiving deducts CoL for all living islands, including critical ones
-func (s *SOMASServer) deductCostOfLiving() error {
+func (s *SOMASServer) deductCostOfLiving(costOfLiving int) {
 	s.logf("start deductCostOfLiving")
 	defer s.logf("finish deductCostOfLiving")
+
 	for id, ci := range s.gameState.ClientInfos {
 		if ci.LifeStatus != shared.Dead {
-			ci.Resources -= config.CostOfLiving
+			ci.Resources -= costOfLiving
```
Comment (preetl) [inline internal/server/turn.go:142]: how come we're not using config.CostOfLiving here? why does it need to be passed into the function?

Comment (lhl2617) [reply internal/server/turn.go:142]: Yep! It's passed in by the caller of this function so we can test this independently of config code (test code shouldn't rely on externally defined constants)

Comment (preetl) [reply internal/server/turn.go:142]: Ah right, makes sense!

**Review by preetl — APPROVED** [review]

**Comment by preetl** [comment]
@lhl2617 really sorry! the conflict was because pr 58 got merged first and I thought I knew how to fix it but these tests are failing now. I don't want to ruin it even more - please help!

**Comment by lhl2617** [comment]
ok sorry i missed this, let me see.

**Pull request merged**